### PR TITLE
Wire up the database/schema

### PR DESCRIPTION
**Pull request by SimonRichardson**
The following wires up the database schema to ensure that we apply the correct patches.

In addition to this, I've removed the sqlite3 testing suite as we're seeing inconsistencies between dqlite and sqlite and people aren't catching the differences.

This fixes this by relocating everything to the right location.

Lastly, now that we've removed the sqlite3 testing suite and forcing the dqlite
one, the foreign keys constraint wasn't being correctly set. We now force it
on every db connection and will fail if it can't be set.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

## QA steps

```sh
$ juju bootstrap lxd test --build-agent
$ juju add-model default
```

### CODE REVIEW

File: agent/agentbootstrap/bootstrap_test.go [hunk 1]
```diff
@@ -451,7 +452,7 @@ func (s *bootstrapSuite) TestInitializeStateFailsSecondTime(c *gc.C) {
 	if err == nil {
 		_ = st.Close()
 	}
-	c.Assert(err, gc.ErrorMatches, "creating controller database schema.*")
+	c.Assert(err, jc.Satisfies, errors.IsAlreadyExists)
```
Comment (tlm) [inline agent/agentbootstrap/bootstrap_test.go:455]: Can we use `errors.Is()` please :)

File: core/database/schema/query_test.go [hunk 2]
```diff
@@ -16,7 +16,7 @@ import (
 )
 
 type querySuite struct {
-	databasetesting.ControllerSuite
+	databasetesting.DqliteSuite
```
Comment (tlm) [inline core/database/schema/query_test.go:19]: Thank you I didn't realise we were using sqlite for the tests. Apologies for assuming what I had written was good to go.

File: core/database/schema/schema_test.go [hunk 3]
```diff
@@ -183,7 +183,7 @@ func (s *schemaSuite) TestEnsureHashBreaks(c *gc.C) {
 	})
 	c.Assert(err, jc.ErrorIsNil)
 
-	schema.Add(MakePatch("SELECT 3"))
+	schema.Add(MakePatch("CREATE TEMP TABLE baz (id INTEGER PRIMARY KEY);"))
```
Comment (tlm) [inline core/database/schema/schema_test.go:186]: Think by this stage we need a `wallyworld` table :)

File: core/watcher/eventsource/package_test.go [hunk 4]
```diff
@@ -22,27 +24,50 @@ func TestPackage(t *testing.T) {
...
+		"core/changestream",
+		"core/database",
+		"core/life",
+		"core/migration",
+		"core/network",
+		"core/resources",
+		"core/secrets",
+		"core/status",
+		"core/watcher",
+		"docker",
```
Comment (tlm) [inline core/watcher/eventsource/package_test.go:45]: This has me curious. Any idea why docker is needed here? Happy to have a look if need be.

File: database/bootstrap.go [hunk 5]
```diff
@@ -91,13 +92,18 @@ func BootstrapDqlite(
 	if err != nil {
 		return errors.Annotatef(err, "opening controller database")
 	}
+
+	if err := pragma.SetPragma(ctx, db, pragma.ForeignKeysPragma, true); err != nil {
+		return errors.Annotate(err, "setting foreign keys pragma")
```
Comment (tlm) [inline database/bootstrap.go:97]: My preference here and I know this style is more preferred by some but to use `fmt.Errorf`. Using annotate still brings cause into play and creates divergent errors paths that can be complicated to use with Is and As.

Not to be militant about it.

File: database/pragma/pragma.go [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+package pragma
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/juju/errors"
+	coredatabase "github.com/juju/juju/core/database"
```
Comment (tlm) [inline database/pragma/pragma.go:12]: Small nit. Should these be seperate blocks to follow rest of Juju import style?

File: database/pragma/pragma.go [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+	ForeignKeysPragma Pragma = "foreign_keys"
+)
+
+// GetPragma returns whether the given pragma is enabled.
+func GetPragma[T any](ctx context.Context, txn coredatabase.TxnRunner, pragam Pragma) (T, error) {
+	var value T
+	err := txn.StdTxn(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		query := fmt.Sprintf("PRAGMA %s", pragam)
+		err := tx.QueryRowContext(ctx, query).Scan(&value)
+		return errors.Trace(err)
```
Comment (tlm) [inline database/pragma/pragma.go:29]: Preference here would be `fmt.Errorf("querying for pragma %q: %w")`

Comment (tlm) [reply database/pragma/pragma.go:29]: Ah I see what you have done below. Disregard.

**Review by tlm — COMMENTED** [review]
Just a few small nits thus far. Will keeping looking at the rest later.

File: core/watcher/eventsource/package_test.go (hunk 4 above)
Comment (SimonRichardson) [reply core/watcher/eventsource/package_test.go:45]: I'm not 100% sure, but we do need to remove it.

File: domain/cloud/state/state.go [hunk 8]
```diff
@@ -620,29 +621,31 @@ func updateRegions(ctx context.Context, tx *sql.Tx, cloudUUID string, regions []
...
-	for _, r := range regions {
-		q := `
+	insertQuery := `
 INSERT INTO cloud_region (uuid, cloud_uuid, name, endpoint, identity_endpoint, storage_endpoint)
 VALUES (?, ?, ?, ?, ?, ?)
 ON CONFLICT(cloud_uuid, name) DO UPDATE SET name=excluded.name,
-                                            endpoint=excluded.endpoint,
-                                            identity_endpoint=excluded.identity_endpoint,
-                                            storage_endpoint=excluded.storage_endpoint`[1:]
+											endpoint=excluded.endpoint,
```
Comment (manadart) [inline domain/cloud/state/state.go:642]: Tabs making this look strange.

File: domain/cloud/state/state.go [hunk 9]
```diff
@@ -221,49 +217,41 @@ func (st *State) UpdateCloudRegionDefaults(
 		return errors.Trace(err)
 	}
 
+	selectStmt := `
+SELECT cloud_region.uuid
```
Comment (manadart) [inline domain/cloud/state/state.go:221]: Vertical alignment.

File: domain/upgrade/state/state.go [hunk 10]
```diff
@@ -200,72 +201,65 @@ func (st *State) SetControllerDone(ctx context.Context, upgradeUUID, controllerI
 	lookForDoneNodesQuery := `
 SELECT (controller_node_id, node_upgrade_completed_at) AS &infoControllerNode.*
 FROM   upgrade_info_controller_node
-WHERE  upgrade_info_uuid = $M.info_uuid`
-	lookForDoneNodesStatement, err := sqlair.Prepare(lookForDoneNodesQuery, infoControllerNode{}, sqlair.M{})
+WHERE  upgrade_info_uuid = $M.info_uuid
+  AND controller_node_id = $M.controller_id;`
```
Comment (jack-w-shaw) [inline domain/upgrade/state/state.go:205]: misaligned AND (or rather, inconsistent)

File: domain/upgrade/state/state.go [hunk 11]
```diff
@@ -200,72 +201,65 @@ func (st *State) SetControllerDone(ctx context.Context, upgradeUUID, controllerI
...
 
 	setNodeToDoneQuery := `
 UPDATE upgrade_info_controller_node
 SET    node_upgrade_completed_at = DATETIME("now")
-WHERE  upgrade_info_uuid = $M.info_uuid 
-       AND controller_node_id = $M.controller_id`
-	setNodeToDoneStatement, err := sqlair.Prepare(setNodeToDoneQuery, sqlair.M{})
+WHERE  upgrade_info_uuid = $M.info_uuid
+       AND controller_node_id = $M.controller_id
+	   AND node_upgrade_completed_at IS NULL;
```
Comment (jack-w-shaw) [inline domain/upgrade/state/state.go:216]: Mix of tabs and spaces on this line

File: domain/upgrade/state/state_test.go [hunk 12]
```diff
@@ -75,13 +81,13 @@ func (s *stateSuite) TestCreateUpgrade(c *gc.C) {
 	c.Assert(err, jc.ErrorIsNil)
 
 	upgradeInfo, nodeInfos := s.getUpgrade(c, s.st, uuid)
-	c.Assert(upgradeInfo, gc.Equals, info{
+	c.Check(upgradeInfo, gc.Equals, info{
```
Comment (manadart) [inline domain/upgrade/state/state_test.go:84]: Good.

File: domain/upgrade/state/state.go [hunk 13]
```diff
@@ -200,72 +201,65 @@ func (st *State) SetControllerDone(ctx context.Context, upgradeUUID, controllerI
...
+		return errors.Annotatef(err, "preparing update node query")
 	}
 
 	completeUpgradeQuery := `
 UPDATE upgrade_info
 SET    completed_at = DATETIME("now")
-WHERE  uuid = $M.info_uuid`
-	completeUpgradeStatement, err := sqlair.Prepare(completeUpgradeQuery, sqlair.M{})
+WHERE  uuid = $M.info_uuid
+AND (
```
Comment (manadart) [inline domain/upgrade/state/state.go:227]: I think you could do this more cogently with a single `WITH` at the top that uses a `HAVING` clause.

**Review by manadart — APPROVED** [review]
Good changes. I like the patch syntax for DB migrations. Just a few formatting things to sort. QA is good.

File: domain/upgrade/state/state.go [hunk 14]
```diff
@@ -200,72 +201,65 @@ func (st *State) SetControllerDone(ctx context.Context, upgradeUUID, controllerI
 	lookForDoneNodesQuery := `
 SELECT (controller_node_id, node_upgrade_completed_at) AS &infoControllerNode.*
 FROM   upgrade_info_controller_node
-WHERE  upgrade_info_uuid = $M.info_uuid`
-	lookForDoneNodesStatement, err := sqlair.Prepare(lookForDoneNodesQuery, infoControllerNode{}, sqlair.M{})
+WHERE  upgrade_info_uuid = $M.info_uuid
+  AND controller_node_id = $M.controller_id;`
+	lookForDoneNodesStmt, err := sqlair.Prepare(lookForDoneNodesQuery, infoControllerNode{}, sqlair.M{})
```
Comment (jack-w-shaw) [inline domain/upgrade/state/state.go:206]: There's still a number of vars suffixed `Statement` in this file. Would be nice to change over the rest whilst you're at it

File: domain/cloud/state/state.go (hunk 8 above)
Comment (SimonRichardson) [reply domain/cloud/state/state.go:642]: My editor hates me.

File: domain/upgrade/state/state.go (hunk 13 above)
Comment (SimonRichardson) [reply domain/upgrade/state/state.go:229]: I did try with `WITH` but I hit a road block. I can look again another time.

**Pull request merged**